smart_open/s3.py

```diff
@@ -190,6 +190,7 @@ def __init__(self, bucket, key, buffer_size=DEFAULT_BUFFER_SIZE,
         self._content_length = self._object.content_length
         self._current_pos = 0
         self._buffer = b''
+        self._buffer_pos = 0
```

Comment:
Thank you for this contribution. I can't help feeling that using a bytestring for _buffer is sub-optimal here. We have to add functionality to output the current position in the buffer, as well as the remainder of the buffer.

Is there a reason why we can't use io.BytesIO? That datatype would encapsulate the above functionality, instead of adding it to the long list of responsibilities that the class already has. More specifically:

- instead of slicing the buffer, we can do _buffer.read(num_bytes)
- _len_remaining_buffer becomes len(_buffer)
- _remaining_buffer becomes _buffer.getvalue()

What do you think?

Reply:
I don't see how that would work, given that io.BytesIO instance's don't have a length.
```py
In [1]: import io

In [2]: buf = io.BytesIO()

In [3]: buf.write(b'foo bar baz')
Out[3]: 11

In [4]: len(buf)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-c8926a5dfb43> in <module>
----> 1 len(buf)

TypeError: object of type '_io.BytesIO' has no len()
```
The only way I can think of calculating the length of what's in a BytesIO buffer is to read everything out of the buffer as a bytestring, which would seem to put us right back where we were before this PR.

Reply:
I do agree with you that it's not ideal for the BufferedInputBase class to be responsible for this. What do you think about creating a ByteStringBuffer class to encapsulate this functionality? That would simplify the implementation of this module's BufferedInputBase, and could also be used in smart_open.http's BufferedInputBase to realize the same performance improvement for files read over HTTP.

Reply:
I do agree with you that it's not ideal for the BufferedInputBase class to be responsible for this. What do you think about creating a ByteStringBuffer class to encapsulate this functionality? That would simplify the implementation of this module's BufferedInputBase, and could also be used in smart_open.http's BufferedInputBase to realize the same performance improvement for files read over HTTP.

If you do like this idea, do you have any opinion on where such a class should be defined? I could put it in the S3 module, but personally I think it makes a little more sense to create a module  just for this class, so it can be used wherever it's needed with less chance of creating an import loop.

Reply:
I like the idea of a class to encapsulate the new responsibility. Let's call it ByteBuffer and put it in bytebuffer.py.

Can you take care of it in this PR?

Reply:
Sounds good. I'll do that, write a few tests for it, and take a shot at making `smart_open.http`s BufferedInputBase class use it for its `_read_buffer`, too.

Should I keep using this PR, or make a new one? If sticking with this PR, do you prefer adding the commits with the new implementation on top of this PR, or force-pushing to replace the current commit? It's all the same to me.

Reply:
Please continue using this PR and add more commits. It'd be good to avoid force-pushing, but it isn't critical.